src/libstd/ffi/c_str.rs

```diff
@@ -704,6 +706,42 @@ impl From<CString> for Box<CStr> {
     }
 }
 
+#[stable(feature = "shared_from_slice2", since = "1.23.0")]
+impl From<CString> for Arc<CStr> {
+    #[inline]
+    fn from(s: CString) -> Arc<CStr> {
+        let arc: Arc<[u8]> = Arc::from(s.into_inner());
+        unsafe { Arc::from_raw(Arc::into_raw(arc) as *const CStr) }
+    }
+}
+
+#[stable(feature = "shared_from_slice2", since = "1.23.0")]
+impl<'a> From<&'a CStr> for Arc<CStr> {
+    #[inline]
+    fn from(s: &CStr) -> Arc<CStr> {
+        let arc: Arc<[u8]> = Arc::from(s.to_bytes_with_nul());
+        unsafe { Arc::from_raw(Arc::into_raw(arc) as *const CStr) }
```

Comment:
Why is this safe? What guarantees that `&'a CStr` outlives the resulting `Arc<CStr>`?

Reply:
Why would `&'a CStr` need to outlive `Arc<CStr>`? `Arc::from` does a new heap allocation and copies over all the contents of `s`.
The `unsafe` block just does `Arc<u8> -> Arc<CStr>`. This is safe since the invariants of `CStr` is maintained.

Reply:
Oh I see. I didn't realize there was a copy here. In hindsight, the `Arc<[u8]>` type should have told me that. And of course this impl: https://doc.rust-lang.org/std/sync/struct.Arc.html#impl-From<&'a [T]>